File: iosApp/iosApp/Presentation/Auth/SignIn/SignInViewModel.swift
//
//  SignInViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/07.
//

import Foundation
import Combine
import Shared

@MainActor
class SignInViewModel: ObservableObject {
    private let signInUseCase: SignInUseCase

    private let signInWithSocialProviderUseCase: SignInWithSocialProviderUseCase

    private let signInWithGoogleUseCase: SignInWithGoogleUseCase

    private let signInWithAppleIdTokenUseCase: SignInWithAppleIdTokenUseCase

    private let signInWithKakaoUseCase: SignInWithKakaoUseCase

    private let completeSignUpForCurrentUserUseCase: CompleteSignUpForCurrentUserUseCase

    @Published private(set) var uiState = SignInUiState.empty

    let event = PassthroughSubject<SignInEvent, Never>()

    private var signInTask: Task<Void, Never>?

    private func signIn(email: String, password: String) {
        uiState.isLoading = true
        uiState.errorMessage = nil

        signInTask?.cancel()
        signInTask = Task {
            do {
                let result = try await signInUseCase.invoke(email: email, password: password)

                if result is AppResultSuccess<AnyObject> {
                    uiState.isLoading = false
                    event.send(.signedIn)
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_invalid_credentials"), table: "Localizable")
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = error.localizedDescription
            }
        }
    }

    private func ensureVisitorAccountCompleted(
        email: String,
        nickname: String,
        signupCompleted: Bool
    ) async -> Bool {
        if signupCompleted {
            return true
        }

        do {
            let result = try await completeSignUpForCurrentUserUseCase.invoke(
                email: email,
                nickname: nickname,
                role: .visitor,
                affiliatedCafeId: nil,
                phoneNumber: nil
            )
            return result is AppResultSuccess<AnyObject>
        } catch {
            return false
        }
    }

    func onAction(_ action: SignInAction) {
        switch action {
        case .emailChanged(let value):
            uiState.email = value
            uiState.errorMessage = nil
        case .passwordChanged(let value):
            uiState.password = value
            uiState.errorMessage = nil
        case .signInTapped:
            signIn(email: uiState.email, password: uiState.password)
        case .socialSignInTapped(let provider):
            uiState.isLoading = true
            uiState.errorMessage = nil
            signInTask?.cancel()
            signInTask = Task {
                if provider == .google {
                    await handleGoogleSignIn()
                    return
                } else if provider == .kakao {
                    await handleKakaoSignIn()
                    return
                }

                do {
                    let result = try await signInWithSocialProviderUseCase.invoke(provider: provider.rawValue)
                    if result is AppResultSuccess<AnyObject> {
                        uiState.isLoading = false
                        event.send(.signedIn)
                    } else {
                        uiState.isLoading = false
                        uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_invalid_credentials"), table: "Localizable")
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.isLoading = false
                    uiState.errorMessage = error.localizedDescription
                }
            }
        case .appleIdTokenReceived(let idToken):
            uiState.isLoading = true
            uiState.errorMessage = nil
            signInTask?.cancel()
            signInTask = Task {
                do {
                    let result = try await signInWithAppleIdTokenUseCase.invoke(idToken: idToken)

                    if let success = result as? AppResultSuccess<AnyObject>,
                       let user = success.data as? Shared.User {
                        let userEmail = String(user.email)
                        let userNickname = String(user.nickname)
                        let isCompleted = await ensureVisitorAccountCompleted(
                            email: userEmail,
                            nickname: userNickname,
                            signupCompleted: user.signupCompleted
                        )
                        if isCompleted {
                            uiState.isLoading = false
                            event.send(.signedIn)
                        } else {
                            uiState.isLoading = false
                            uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_apple_failed"), table: "Localizable")
                        }
                    } else {
                        uiState.isLoading = false
                        uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_apple_failed"), table: "Localizable")
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.isLoading = false
                    uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_apple_failed"), table: "Localizable")
                }
            }
        }
    }

    init(
        signInUseCase: SignInUseCase = KoinInitializerKt.resolveSignInUseCase(),
        signInWithGoogleUseCase: SignInWithGoogleUseCase = KoinInitializerKt.resolveSignInWithGoogleUseCase(),
        signInWithAppleIdTokenUseCase: SignInWithAppleIdTokenUseCase = KoinInitializerKt.resolveSignInWithAppleIdTokenUseCase(),
        signInWithKakaoUseCase: SignInWithKakaoUseCase = KoinInitializerKt.resolveSignInWithKakaoUseCase(),
        completeSignUpForCurrentUserUseCase: CompleteSignUpForCurrentUserUseCase = KoinInitializerKt.resolveCompleteSignUpForCurrentUserUseCase()
    ) {
        self.signInUseCase = signInUseCase
        self.signInWithSocialProviderUseCase = SignInWithSocialProviderUseCase(signInUseCase: signInUseCase)
        self.signInWithGoogleUseCase = signInWithGoogleUseCase
        self.signInWithAppleIdTokenUseCase = signInWithAppleIdTokenUseCase
        self.signInWithKakaoUseCase = signInWithKakaoUseCase
        self.completeSignUpForCurrentUserUseCase = completeSignUpForCurrentUserUseCase
    }

    deinit {
        signInTask?.cancel()
    }

    private func handleGoogleSignIn() async {
        do {
            let result = try await signInWithGoogleUseCase.invoke()

            if let success = result as? AppResultSuccess<AnyObject>,
               let user = success.data as? Shared.User {
                let isCompleted = await ensureVisitorAccountCompleted(
                    email: String(user.email),
                    nickname: String(user.nickname),
                    signupCompleted: user.signupCompleted
                )
                if isCompleted {
                    uiState.isLoading = false
                    event.send(.signedIn)
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_google_failed"), table: "Localizable")
                }
            } else {
                uiState.isLoading = false
                uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_google_failed"), table: "Localizable")
            }
        } catch {
            if Task.isCancelled { return }
            uiState.isLoading = false
            uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_google_failed"), table: "Localizable")
        }
    }

    private func handleKakaoSignIn() async {
        do {
            let result = try await signInWithKakaoUseCase.invoke()

            if let success = result as? AppResultSuccess<AnyObject>,
               let user = success.data as? Shared.User {
                let userNickname = String(user.nickname)
                let completionNickname = userNickname.isEmpty
                    ? String(localized: String.LocalizationValue("signin_default_kakao_nickname"), table: "Localizable")
                    : userNickname

                if await ensureVisitorAccountCompleted(
                    email: String(user.email),
                    nickname: completionNickname,
                    signupCompleted: user.signupCompleted
                ) {
                    uiState.isLoading = false
                    event.send(.signedIn)
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_kakao_failed"), table: "Localizable")
                }
            } else {
                uiState.isLoading = false
                uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_kakao_failed"), table: "Localizable")
            }
        } catch {
            if Task.isCancelled { return }
            uiState.isLoading = false
            uiState.errorMessage = String(localized: String.LocalizationValue("signin_error_kakao_failed"), table: "Localizable")
        }
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/auth/signin/SignInViewModel.kt
package com.hhp227.concafe.presentation.auth.signin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import concafe.composeapp.generated.resources.Res
import concafe.composeapp.generated.resources.signin_default_kakao_nickname
import concafe.composeapp.generated.resources.signin_error_apple_ios_only
import concafe.composeapp.generated.resources.signin_error_google_failed
import concafe.composeapp.generated.resources.signin_error_invalid_credentials
import concafe.composeapp.generated.resources.signin_error_kakao_failed
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.UserRole
import com.hhp227.concafe.domain.usecase.CompleteSignUpForCurrentUserUseCase
import com.hhp227.concafe.domain.usecase.SignInUseCase
import com.hhp227.concafe.domain.usecase.SignInWithGoogleUseCase
import com.hhp227.concafe.domain.usecase.SignInWithKakaoUseCase
import org.jetbrains.compose.resources.getString

class SignInViewModel(
    private val signInUseCase: SignInUseCase,
    private val signInWithGoogleUseCase: SignInWithGoogleUseCase,
    private val signInWithKakaoUseCase: SignInWithKakaoUseCase,
    private val completeSignUpForCurrentUserUseCase: CompleteSignUpForCurrentUserUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(SignInUiState.empty())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<SignInEvent>()
    val event = _event.asSharedFlow()

    private fun changeEmail(value: String) {
        _uiState.update { it.copy(email = value, errorMessage = null) }
    }

    private fun changePassword(value: String) {
        _uiState.update { it.copy(password = value, errorMessage = null) }
    }

    private fun signIn(email: String, password: String) {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }

        viewModelScope.launch {
            when (val result = signInUseCase.invoke(email, password)) {
                is AppResult.Success -> {
                    _uiState.update { it.copy(isLoading = false, errorMessage = null) }
                    _event.emit(SignInEvent.SignedIn)
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = getString(Res.string.signin_error_invalid_credentials)
                        )
                    }
                }
            }
        }
    }

    private suspend fun ensureVisitorAccountCompleted(
        email: String,
        nickname: String,
        signupCompleted: Boolean
    ): Boolean {
        if (signupCompleted) return true
        return when (
            completeSignUpForCurrentUserUseCase.invoke(
                email = email,
                nickname = nickname,
                role = UserRole.VISITOR
            )
        ) {
            is AppResult.Success -> true
            is AppResult.Failure -> false
        }
    }

    fun onAction(action: SignInAction) {
        when (action) {
            is SignInAction.ChangeEmail -> changeEmail(action.value)
            is SignInAction.ChangePassword -> changePassword(action.value)
            SignInAction.ClickEmailSignIn -> signIn(uiState.value.email, uiState.value.password)
            is SignInAction.ClickSocialSignIn -> {
                _uiState.update { it.copy(isLoading = true, errorMessage = null) }
                viewModelScope.launch {
                    when (action.provider) {
                        SignInProvider.GOOGLE -> {
                            when (val result = signInWithGoogleUseCase.invoke()) {
                                is AppResult.Success -> {
                                    if (ensureVisitorAccountCompleted(
                                            email = result.data.email,
                                            nickname = result.data.nickname,
                                            signupCompleted = result.data.signupCompleted
                                        )
                                    ) {
                                        _uiState.update { it.copy(isLoading = false, errorMessage = null) }
                                        _event.emit(SignInEvent.SignedIn)
                                    } else {
                                        _uiState.update {
                                            it.copy(
                                                isLoading = false,
                                                errorMessage = getString(Res.string.signin_error_google_failed)
                                            )
                                        }
                                    }
                                }
                                is AppResult.Failure -> {
                                    _uiState.update {
                                        it.copy(
                                            isLoading = false,
                                            errorMessage = getString(Res.string.signin_error_google_failed)
                                        )
                                    }
                                }
                            }
                        }
                        SignInProvider.APPLE -> {
                            _uiState.update {
                                it.copy(
                                    isLoading = false,
                                    errorMessage = getString(Res.string.signin_error_apple_ios_only)
                                )
                            }
                        }
                        SignInProvider.KAKAO -> {
                            when (val result = signInWithKakaoUseCase.invoke()) {
                                is AppResult.Success -> {
                                    if (ensureVisitorAccountCompleted(
                                            email = result.data.email,
                                            nickname = result.data.nickname.ifBlank {
                                                getString(Res.string.signin_default_kakao_nickname)
                                            },
                                            signupCompleted = result.data.signupCompleted
                                        )
                                    ) {
                                        _uiState.update { it.copy(isLoading = false, errorMessage = null) }
                                        _event.emit(SignInEvent.SignedIn)
                                    } else {
                                        _uiState.update {
                                            it.copy(
                                                isLoading = false,
                                                errorMessage = getString(Res.string.signin_error_kakao_failed)
                                            )
                                        }
                                    }
                                }
                                is AppResult.Failure -> {
                                    _uiState.update {
                                        it.copy(
                                            isLoading = false,
                                            errorMessage = getString(Res.string.signin_error_kakao_failed)
                                        )
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
